### PR TITLE
[KBFS-2719] Initial KBFS Navigation Barebones

### PR DESCRIPTION
**Pull request by jzila**
Just starting out the initial KBFS root structure. Navigation through the tree is next.

cc @keybase/kbfs-hackers

### CODE REVIEW

File: shared/constants/types/fs.js [hunk 1]
```diff
@@ -6,3 +6,21 @@ export type _State = {
 }
 
 export type State = I.RecordOf<_State>
+
+export type FolderVisibility = 'private' | 'public' | 'team' | null
+
+export type FolderHeaderProps = {
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: so typically constants/types contains things related to the store only. So i imagine we will have FolderVisibiilty as metadata on a path, but FileRowProps is just props for a single row and likely belongs with the dumb component (fs/rows/file.js or whatever)

File: shared/constants/types/fs.js [hunk 2]
```diff
@@ -6,3 +6,21 @@ export type _State = {
...
+export type FileRowProps = {
+  path: string,
+  icon: IconType,
+  showFileData: () => void,
+}
+
+export type FolderProps = {
+  path: string,
+  visibility: FolderVisibility,
+  items: Array<string>,
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: Just a heads up flow does have opaque types which we're starting to use. This allows you to have something that is effectively a string but it treated as an independent type. This allows you to ensure function that take a 'Path' will only pass if a 'Path' is passed and not some arbitrary string.
We don't need to add this complexity now but just pointing out a direction we're going

Comment (chrisnojima) [reply shared/constants/types/fs.js]: for example https://github.com/keybase/client/blob/41fc5de285d2708841a36012ea8b519bb8048b1b/shared/constants/types/devices.js#L5 . deviceIDs are strings, but we want to treat them as not-strings to get stricter enforcement of types

File: shared/fs/index.js [hunk 3]
```diff
@@ -1,30 +1,110 @@
...
-import * as React from 'react'
-import {globalStyles, globalMargins} from '../styles'
-import {Box, Button, Text, ButtonBar} from '../common-adapters'
-
-type Props = {
-  counter: number,
-  you: ?string,
-  increase: () => void,
-  increase10: () => void,
+import React, {PureComponent} from 'react'
```
Comment (chrisnojima) [inline shared/fs/index.js]: we're moving towards using import * as React and using extends React.PureCompoent below

File: shared/fs/index.js [hunk 4]
```diff
@@ -1,30 +1,110 @@
...
+import {FolderVisibility} from '../constants/types/fs'
```
Comment (chrisnojima) [inline shared/fs/index.js]: typically `import * as Types`

File: shared/fs/index.js [hunk 5]
```diff
@@ -1,30 +1,110 @@
...
+import {globalStyles, globalColors, globalMargins} from '../styles'
+import {isMobile} from '../constants/platform'
```
Comment (chrisnojima) [inline shared/fs/index.js]: isMobile was pulled into style so you can put it into the line above

File: shared/fs/index.js [hunk 6]
```diff
@@ -1,30 +1,110 @@
...
 
-const containerStyle = {
+const stylesContainer = {
   ...globalStyles.flexBoxColumn,
+  ...globalStyles.fullHeight,
   flex: 1,
-  padding: globalMargins.xlarge,
 }
 
+const rootFolders = [
```
Comment (chrisnojima) [inline shared/fs/index.js]: So this is great for the skeleton and keeping things simple. Just a heads up that I would assume the actual implementation would have connected rows and the props that would be passed to the children would be mostly ids (or paths or whatever the identifier is)

data like rootFolders would typically be derived from the store and created by the connect wrapper

File: shared/fs/index.js [hunk 7]
```diff
@@ -1,30 +1,110 @@
...
+
+type FolderProps = {
+  path: string,
+  visibility: FolderVisibility,
+  items: Array<string>,
+}
+
+const FolderHeader = ({title}: FolderHeaderProps) => (
+  <Box>
+    <Box style={{...stylesCommonRow, alignItems: 'center', borderBottomWidth: 0}}>
```
Comment (chrisnojima) [inline shared/fs/index.js]: This style is fine and we do it everywhere, just a heads up if you have items that are mutating a bunch (likely not in this case) when this render happens it'll be making this style object over and over and it'll fail any equality comparison so it'll force a re-render. This isn't a problem here as the connect call is pure and will not rerender if nothing is changing, but if we did have more stuff rendering this style could cause render thrashing.

tldr; {foo: 'bar'} !== {foo: 'bar'}

File: shared/fs/index.js [hunk 8]
```diff
@@ -1,30 +1,110 @@
...
+      </Text>
+    </Box>
+  </Box>
+)
+
+
+const FileRow = ({path, icon, showFileData}: FileRowProps) => (
+  <ClickableBox onClick={showFileData} style={{...stylesCommonRow}}>
+    <Box key={path} style={{...globalStyles.flexBoxRow, alignItems: 'center', flex: 1}}>
+      <Icon type={icon} style={{marginRight: 16}} />
```
Comment (chrisnojima) [inline shared/fs/index.js]: we do have globalMargins for these types of consts

Comment (chrisnojima) [reply shared/fs/index.js]: https://github.com/keybase/client/blob/41fc5de285d2708841a36012ea8b519bb8048b1b/shared/styles/index.desktop.js#L55

File: shared/fs/index.js [hunk 9]
```diff
@@ -1,30 +1,110 @@
...
+      <Box style={{...globalStyles.flexBoxColumn, flex: 1, justifyContent: 'flex-start'}}>
+        <Text type="Body" style={{flex: 1}}>
```
Comment (chrisnojima) [inline shared/fs/index.js]: unclear if flex:1 is what you want here. maybe width: 100% or the parent doing justifyContent: stretch. but this is all placeholder so its fine

File: shared/fs/index.js [hunk 10]
```diff
@@ -1,30 +1,110 @@
...
+          {path}
+        </Text>
+      </Box>
+    </Box>
+  </ClickableBox>
+)
+
+class Folder extends PureComponent<FolderProps> {
+  _renderRow(index, item) {
```
Comment (chrisnojima) [inline shared/fs/index.js]: Just want to point out a subtlety about member methods. This works fine in this instance but note this is *not* a bound method. 

aka. if this method referred to `this.myValue` the 'this' pointer could be undefined.

aka
```
_renderRow(index, item) {
  console.log(this.myValue); 
}
```

will work if you do
```
const folder = new Folder(...)
folder._renderRow(0, 0)
```

but if you do
```
const folder = new Folder(...)
const whoops = folder._renderRow
whoops(0, 0)
```

that will fail. While this seems like a contrived example this is exactly what we're doing passing this._renderRow to `<List items={items} renderItem={this._renderRow} />`

There are a couple of solutions to this (bind, etc) but the simplest and best react wise is to instead make the member an arrow function so 'this' is closed over and automatically bound

tldr;

```
_renderRow = (index, item) => {
```

does not have this problem

File: shared/fs/index.js [hunk 11]
```diff
@@ -1,30 +1,110 @@
...
+      <FileRow path={item.path} icon={item.icon} showFileData={item.showFileData} />
+    )
+  }
+
+  render() {
+    const iconTypes : IconType = {
+      folder: isMobile ? 'icon-folder-private-24' : 'icon-folder-private-24',
+      file: isMobile ? 'icon-file-24' : 'icon-file-24',
+    }
+    const items = [
```
Comment (chrisnojima) [inline shared/fs/index.js]: This form will cause a lot of render thrashing as whats passed down to <List will be new every time (even if nothing changes). Ideally the connect would return just an array of ids thats passed down and each row would be connected. 
Since this is a simple skeleton its fine to not have this be connected right now, but still you likely want to just pass this.props.items into <List directly and have _renderRow do this kind of map (the item passed there is essentially what you get on line 88 here)

**Comment by chrisnojima** [comment]
Thanks for this PR! I'm about to add a bunch of feedback. Since this team is just ramping up I'm adding a bunch of feedback. We can keep this as-is, or you can opt to make some of the changes. I just want to point out ways in which I see this code moving in the future. The short term goal is to get a little skeleton out so keeping things simple is ideal.

File: shared/fs/index.js (hunk 7 above)
Comment (jzila) [reply shared/fs/index.js]: So what's the right answer here?

File: shared/fs/index.js (hunk 9 above)
Comment (jzila) [reply shared/fs/index.js]: I see `flex: 1` everywhere...from my reading this seems to be equivalent to `width: 100%`, but I'm not certain.

File: shared/fs/index.js (hunk 7 above)
Comment (chrisnojima) [reply shared/fs/index.js]: just pull it into its own variable. below just make a const FolderBoxStyle = {...stylesCommonRow, ...} etc

File: shared/fs/index.js (hunk 9 above)
Comment (chrisnojima) [reply shared/fs/index.js]: flex: 1 will have it grow to fill the parent. you really only intend that to happen for the width (it's a row)

File: shared/fs/index.js (hunk 7 above)
Comment (jzila) [reply shared/fs/index.js]: Does this apply to all the `style={{foo}}` attributes?

Comment (chrisnojima) [reply shared/fs/index.js]: it applies to all props period

Comment (jzila) [reply shared/fs/index.js]: We seem to have this issue on the Devices screen then, where we have tons of `style={{foo}}`

Comment (chrisnojima) [reply shared/fs/index.js]: this is why i mentioned if the things mutate a bunch. on the devices screen they never mutate so it renders once on mount

File: shared/fs/index.js (hunk 10 above)
Comment (songgao) [reply shared/fs/index.js]: Does it mean you could do `whoops(folder, 0, 0)` which would be equivalent to `folder._renderRow(0,0)`?

**Comment by jzila** [comment]
@chrisnojima: I think this is ready for another look. LMK if there are draconian changes needed, otherwise I'll start creating tickets to parallelize work.

File: shared/constants/types/fs.js [hunk 12]
```diff
@@ -2,7 +2,27 @@
...
 
 export type _State = {
-  counter: number,
+  path: Path,
+  pathItems: I.Map<Path, PathItem>,
 }
 
 export type State = I.RecordOf<_State>
+
+export type PathItem = {
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: this type should be another record actually. so i'd rename this to _PathItem and have a type PathItem = I.RecordOf<_PathItem>.
Typically with immutable we want everything in the hierarchy to be immutable else you'll run into issue.

sidebar: one reason you want immutable with redux is that it helps ensure sub items that haven't changed will be === (their hash trie thing is pretty cool). We also want the opposite thing, which is to say if a prop in an object changes we want the parent object to actually change so mapStateToProps (etc) will equality fail and cause a rerender.

for example
const obj = {foo: 'bar'}
const obj2 = obj
obj.foo = 'whut'
obj === obj2 // whoops

So this can cause you to update the data and yet nothing renders. As a side effect its possible the map gets updated or something so you might be masking the issue but it really should be avoided.

The other issue is if you use immutable's helpers they mostly assume its a hierarchy of immutable objects. aka
myObject.getIn(['this', 'is', 'a path', 'of', 'properties','that','are','nested'])

so if these are all maps of maps of records of maps of records, etc it'll work fine. if its maps of objects of maps of records of objects it *could* fail.

easier if its all immutable top to bottom

File: shared/constants/types/fs.js [hunk 13]
```diff
@@ -2,7 +2,27 @@
...
+  type: PathType,
+  children: I.List<string> | null,
+}
+
+export type Visibility = 'private' | 'public' | 'team' | null
+
+export type PathType = 'folder' | 'file' | 'symlink' | 'exec'
+
+export opaque type Path = string | null
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: ?string is likely better

File: shared/constants/types/fs.js [hunk 14]
```diff
@@ -2,7 +2,27 @@
...
 }
 
 export type State = I.RecordOf<_State>
+
+export type PathItem = {
+  type: PathType,
+  children: I.List<string> | null,
+}
+
+export type Visibility = 'private' | 'public' | 'team' | null
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: why allow null visibility

File: shared/constants/types/fs.js [hunk 15]
```diff
@@ -2,7 +2,27 @@
...
-  counter: number,
+  path: Path,
+  pathItems: I.Map<Path, PathItem>,
 }
 
 export type State = I.RecordOf<_State>
+
+export type PathItem = {
+  type: PathType,
+  children: I.List<string> | null,
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: you want a disjoint union here and stronger typing. aka i *think* the intention is something like
```
type _PathItem = 
| {
  type: 'file'
}  
| {
type: 'folder' | 'symlink' | 'exec',
children: I.List<Path>
}
```

File: shared/constants/types/fs.js [hunk 16]
```diff
@@ -2,7 +2,27 @@
...
+}
+
+export type Visibility = 'private' | 'public' | 'team' | null
+
+export type PathType = 'folder' | 'file' | 'symlink' | 'exec'
```
Comment (chrisnojima) [inline shared/constants/types/fs.js]: what is symlink / exec used for. if you want to draw the children in a list as symlinks maybe your list needs to be something other than string. 
aka
```
// children: I.List<string> , instead
children: I.List<ChildType>

type ChildType = {
  type: PathType,
  path: Path
}
```

File: shared/fs/container.js [hunk 17]
```diff
@@ -1,16 +1,57 @@
...
 import * as FSGen from '../actions/fs-gen'
-import Fs from '.'
+import Files from '.'
+import {navigateAppend, navigateUp} from '../actions/route-tree'
+import * as Types from '../constants/types/fs'
 
-const mapStateToProps = (state: TypedState) => ({
-  counter: state.fs.counter,
-  you: state.config.username,
+type StateProps = {
```
Comment (chrisnojima) [inline shared/fs/container.js]: i don't think we actually need these types to be explicit in the return values. flowtyping of connect has gotten a lot better and it should be inferable

File: shared/fs/container.js [hunk 18]
```diff
@@ -1,16 +1,57 @@
...
+  items: I.List<string>,
+  pathItems: I.Map<Path, PathItem>,
+}
+
+type DispatchProps = {
+  onViewFolder: (path: Types.Path) => void,
+}
+
+const mapStateToProps = (state: TypedState, ownProps): StateProps => {
+  var path = state.fs.defaultPath
```
Comment (chrisnojima) [inline shared/fs/container.js]: This likely shouldn't be in the store. This is really a const that'll never mutate.

File: shared/fs/container.js [hunk 19]
```diff
@@ -1,16 +1,57 @@
...
+  if (ownProps.routeProps) {
```
Comment (chrisnojima) [inline shared/fs/container.js]: you will always have routeProps unless you're embedding this else where (likely not happening). so better to just simplify and remove it

File: shared/fs/container.js (hunk 18 above)
Comment (chrisnojima) [inline shared/fs/container.js]: we don'' use var ever, use let instead if you need to but strive to use const

File: shared/fs/container.js [hunk 20]
```diff
@@ -1,16 +1,57 @@
...
+    path = ownProps.routeProps.get('path', path)
```
Comment (chrisnojima) [inline shared/fs/container.js]: really could be
```
const path = ownProps.routeProps.get('path', '/')
```
 or the '/' could be in Constants.root or something

File: shared/fs/container.js [hunk 21]
```diff
@@ -1,16 +1,57 @@
...
 })
 
-const mapDispatchToProps = (dispatch: Dispatch) => ({
-  increase: () => dispatch(FSGen.createIncreaseCount({amount: 1})),
-  increase10: () => dispatch(FSGen.createIncreaseCount({amount: 10})),
+const mergeProps = (stateProps: StateProps, dispatchProps: DispatchProps) => ({
+  name: Types.getPathName(stateProps.path),
+  path: stateProps.path,
+  visibility: Types.getPathVisibility(stateProps.path),
+  items: stateProps.items.map(name => {
```
Comment (chrisnojima) [inline shared/fs/container.js]: it may make sense for this to actually be the representation of children in the tree. we should discuss

File: shared/fs/index.js [hunk 22]
```diff
@@ -1,30 +1,119 @@
...
+
+const iconTypes : IconType = {
+  folder: isMobile ? 'icon-folder-private-24' : 'icon-folder-private-24',
+  file: isMobile ? 'icon-file-24' : 'icon-file-24',
+}
+
+export class Files extends React.PureComponent<FolderProps> {
+  _renderRow = (index, item) => {
+    const iconType = iconTypes[item.type]
+    const showFileData = () => {
```
Comment (chrisnojima) [inline shared/fs/index.js]: we should avoid creating closures like this in what is essentially a render function. Doing this will cause the showfileData prop to change on every render causing further rerenders down the line. Likely better is to keep it accepting a param and have FileRow call it with the path itself
aka
```
showFileData={item.type === 'folder' ? this.props.onViewFolder : this.props.onViewFile}
```
```

File: shared/fs/index.js [hunk 23]
```diff
@@ -1,30 +1,119 @@
...
+      if (item.type == 'folder') {
```
Comment (chrisnojima) [inline shared/fs/index.js]: === always. do you have linting turned on?

File: shared/fs/index.js [hunk 24]
```diff
@@ -1,30 +1,119 @@
...
+    return (
+      <FileRow name={item.name} path={item.path} icon={iconType} showFileData={showFileData} />
+    )
+  }
+
+  render() {
+    return (
+      <Box style={styleOuterContainer}>
+        <Box style={stylesContainer}>
+          {this.props.onBack && this.props.path !== '/keybase' && (
```
Comment (chrisnojima) [inline shared/fs/index.js]: better to make this just check onBack and have that be null if the path is the root. dumb component shouldn't know this detail and semantically we just want a back button when onBack exists

File: shared/constants/types/fs.js (hunk 15 above)
Comment (jzila) [reply shared/constants/types/fs.js]: Yeah I thought of doing that but wasn't sure how much to trust Flow (given its failure to discover that my `FolderProps` were getting an Immutable object instead of an array).

File: shared/fs/index.js [hunk 25]
```diff
@@ -1,30 +1,119 @@
...
+type FolderProps = {
+  name: string,
+  path: Types.Path,
+  visibility: Types.Visibility,
+  items: Array<string>,
+  onViewFolder: (p: Path) => void,
+  onViewFile: (p: Path) => void,
+}
+
+const FolderBoxStyle = {...globalStyles.flexBoxColumn, flex: 1, justifyContent: 'stretch'}
```
Comment (songgao) [inline shared/fs/index.js]: The naming scheme seems different. Did you mean `stylesFolderBox`?

File: shared/fs/index.js (hunk 7 above)
Comment (songgao) [reply shared/fs/index.js]: I think this would still cause re-rendering since `style` is a new object every time. Did you intend to do this because `FolderHeader` rarely gets re-rendered, or is this supposed to be `FolderBoxStyle` as well?

Comment (jzila) [reply shared/fs/index.js]: Yeah the header will never be re-rendered.

File: shared/fs/index.js (hunk 25 above)
Comment (jzila) [reply shared/fs/index.js]: Hah yeah I wrote that first, so I hadn't yet adapted the naming conventions of the codebase.

**Comment by jzila** [comment]
This is still going to fail Flow, but I'll work on fixing that tomorrow.

**Comment by jzila** [comment]
@chrisnojima: one flow error left that has me completely stumped. Otherwise ready for another look.

**Comment by jzila** [comment]
@chrisnojima: Ready for thumb, or would you still recommend some changes?

**Comment by chrisnojima** [comment]
but lets not merge until we get this mobile app out
